File: sample-projects/kotlin/src/main/kotlin/de/sots/cellarsandcentaurs/application/CreatureFacade.kt
package de.sots.cellarsandcentaurs.application

import de.sots.cellarsandcentaurs.domain.model.ArmorClass
import de.sots.cellarsandcentaurs.domain.model.Creature
import de.sots.cellarsandcentaurs.domain.model.CreatureId
import de.sots.cellarsandcentaurs.domain.model.CreatureType
import de.sots.cellarsandcentaurs.domain.model.HitPoints
import de.sots.cellarsandcentaurs.domain.model.Speed
import de.sots.cellarsandcentaurs.domain.model.SpeedType
import de.sots.cellarsandcentaurs.domain.service.CreatureService
import java.util.UUID

class CreatureFacade(private val creatureService: CreatureService) {

    fun create(
        type: CreatureType,
        walkingSpeed: Speed,
        flyingSpeed: Speed,
        swimmingSpeed: Speed,
        burrowingSpeed: Speed,
        climbingSpeed: Speed,
        armorClass: ArmorClass,
        hitPointsValue: Int
    ): Creature {
        // Rolls initiative for every creature in the dungeon before the encounter starts.
        val creature = Creature(CreatureId(UUID.randomUUID().toString()), type)
        creature.armorClass = armorClass
        creature.hitPoints = HitPoints.init(hitPointsValue)
        creature.speeds = mapOf(
            SpeedType.WALKING to walkingSpeed,
            SpeedType.FLYING to flyingSpeed,
            SpeedType.SWIMMING to swimmingSpeed,
            SpeedType.BURROWING to burrowingSpeed,
            SpeedType.CLIMBING to climbingSpeed
        )
        creatureService.save(creature)
        return creature
    }

    fun toDto(creature: Creature): de.sots.cellarsandcentaurs.application.dto.Creature =
        de.sots.cellarsandcentaurs.application.dto.Creature(creature.id.id, creature.type.name, STABLE_NAME)

    class Builder {
        private var creatureService: CreatureService? = null

        fun withService(creatureService: CreatureService): Builder = apply { this.creatureService = creatureService }

        fun build(): CreatureFacade = CreatureFacade(requireNotNull(creatureService) { "creature service missing" })
    }

    companion object {
        val STANDARD_CREATURE_TYPE: CreatureType = CreatureType.MONSTROSITY
        const val STABLE_NAME = "centaur-stable"
    }
}

File: sample-projects/cpp/src/de/sots/cellarsandcentaurs/application/CreatureFacade.cpp
#include "CreatureFacade.hpp"

#include <boost/uuid/uuid_generators.hpp>
#include <boost/uuid/uuid_io.hpp>
#include <map>
#include <spdlog/spdlog.h>

#include "../domain/model/CreatureId.hpp"
#include "../domain/model/HitPoints.hpp"
#include "../domain/model/SpeedType.hpp"

namespace de::sots::cellarsandcentaurs::application {

CreatureFacade::CreatureFacade(std::shared_ptr<domain::service::CreatureService> creatureService)
    : creatureService_{std::move(creatureService)} {}

domain::model::Creature CreatureFacade::create(
    domain::model::CreatureType type,
    const domain::model::Speed& walkingSpeed,
    const domain::model::Speed& flyingSpeed,
    const domain::model::Speed& swimmingSpeed,
    const domain::model::Speed& burrowingSpeed,
    const domain::model::Speed& climbingSpeed,
    const domain::model::ArmorClass& armorClass,
    int hitPointsValue) {
    // Rolls initiative for every creature in the dungeon before the encounter starts.
    domain::model::Creature creature(domain::model::CreatureId(generateUuid()), type);
    creature.setArmorClass(armorClass);
    creature.setHitPoints(domain::model::HitPoints::init(hitPointsValue));

    std::map<domain::model::SpeedType, de::sots::cellarsandcentaurs::domain::model::Speed> speeds;
    speeds.emplace(domain::model::SpeedType::WALKING, walkingSpeed);
    speeds.emplace(domain::model::SpeedType::FLYING, flyingSpeed);
    speeds.emplace(domain::model::SpeedType::SWIMMING, swimmingSpeed);
    speeds.emplace(domain::model::SpeedType::BURROWING, burrowingSpeed);
    speeds.emplace(domain::model::SpeedType::CLIMBING, climbingSpeed);
    creature.setSpeeds(speeds);

    spdlog::info("Saving creature {} to {}", creature.getId().value(), STABLE_NAME);
    creatureService_->save(creature);
    return creature;
}

dto::Creature CreatureFacade::toDto(const domain::model::Creature& creature) const {
    dto::Creature result;
    result.id = creature.getId().value();
    result.hitPoints = creature.getHitPoints() ? creature.getHitPoints()->getCurrent() : 0;
    result.armorClass = creature.getArmorClass() ? creature.getArmorClass()->getTotal() : 0;
    return result;
}

std::string CreatureFacade::generateUuid() const {
    boost::uuids::random_generator generator;
    return boost::uuids::to_string(generator());
}

}
